File: relay_lab/web/faders.js
'use strict';
const FaderUI=(()=>{
  const types=['short','medium','long'];
  const ids=['fader-duration','fader-cap','fader-ceiling','fader-interval'];
  let timer=null,pending=null,sending=false,jobId=null,version=0;
  const isMode=()=>['account-test','gateway-test'].includes(mode)&&$('load-mode').value==='faders';
  const values=()=>types.map(p=>Number($('fader-'+p+'-number').value));
  const liveJob=()=>currentJob?.faders&&currentJob.id===active&&currentJob.status==='running'?currentJob:null;
  function paint(){
    const maximum=Math.max(1,Number($('fader-range').value)||20,...values());$('fader-range').value=maximum;
    types.forEach(p=>{const slider=$('fader-'+p);slider.max=maximum;slider.value=$('fader-'+p+'-number').value;slider.style.setProperty('--fader-fill',(Number(slider.value)/maximum*100)+'%');slider.setAttribute('aria-valuetext',slider.value+' 条目标在途请求');$('fader-'+p+'-max').textContent=maximum;});
    $('fader-target-total').textContent=number(values().reduce((a,b)=>a+b,0));
  }
  function config(){return {targets:values(),output_limits:types.map(p=>Number($('fader-'+p+'-limit').value)),duration:Number($('fader-duration').value),max_requests:Number($('fader-cap').value),max_inflight:Number($('fader-ceiling').value),refill_interval:Number($('fader-interval').value)};}
  function sync(){
    const on=isMode(),live=liveJob();$('fader-fields').hidden=!on;$('fader-panel').hidden=!(on||currentJob?.faders);
    ids.forEach(id=>$(id).disabled=!on||Boolean(active));
    types.forEach(p=>{const locked=Boolean(active)&&!(live&&live.faders.accepting);$('fader-'+p).disabled=locked;$('fader-'+p+'-number').disabled=locked;$('fader-'+p+'-limit').disabled=Boolean(active)||submitting;});
    $('fader-pause').hidden=!live;$('fader-pause').disabled=Boolean(live&&!live.faders.accepting);
    if(!active&&!currentJob?.faders)$('fader-control-state').textContent='启动后可实时调节；目标为 0 时不发请求';
    paint();
  }
  async function send(){
    if(sending)return;sending=true;
    try{
      while(pending){
        const live=liveJob();if(!live){pending=null;break;}
        const update=pending;pending=null;
        const job=await api('/api/jobs/'+live.id+'/faders',update);
        if(selected===job.id)renderJob(job);
      }
      $('fader-control-error').hidden=true;
    }catch(e){pending=null;$('fader-control-error').textContent=e.message;$('fader-control-error').hidden=false;}
    finally{sending=false;if(currentJob)render(currentJob);}
  }
  function queue(update){pending={...(pending||{}),...update};if(timer)clearTimeout(timer);timer=setTimeout(send,200);}
  function init(){
    types.forEach(p=>{
      $('fader-'+p).addEventListener('input',()=>{$('fader-'+p+'-number').value=$('fader-'+p).value;paint();if(liveJob())queue({targets:values()});});
      $('fader-'+p+'-number').addEventListener('change',()=>{paint();if(liveJob())queue({targets:values()});});
    });
    $('fader-range').addEventListener('change',paint);
    $('fader-pause').addEventListener('click',()=>{const live=liveJob();if(live)queue({paused:!live.faders.paused});});
    paint();
  }
  function render(job){
    const f=job.faders;if(!f){sync();return;}
    if(jobId!==job.id){jobId=job.id;version=0;}
    if(f.version<version)return;version=f.version;
    if(!pending&&!sending&&job.status==='running')types.forEach((p,i)=>{const input=$('fader-'+p+'-number');if(document.activeElement!==input)input.value=f.targets[i];});
    const editable=job.id===active&&job.status==='running'&&f.accepting;
    $('fader-control-state').textContent=editable?(f.paused?'已暂停补发；在途请求自然结束':`运行中 · 每 ${f.refill_interval} 秒补足目标`):job.status==='running'?'已到运行上限，正在收尾':'记录已结束；推子值可用于新测试';
    $('fader-pause').textContent=f.paused?'继续补发':'暂停补发';
    $('fader-issued').textContent=number(f.issued_requests);
    $('fader-inflight').textContent=number(f.inflight);
    $('fader-waiting').textContent=number(f.channels.reduce((n,c)=>n+c.waiting,0));
    $('fader-receiving').textContent=number(f.channels.reduce((n,c)=>n+c.receiving,0));
    for(const c of f.channels){$('fader-'+c.profile+'-state').textContent=`等待 ${c.waiting} · 接收中 ${c.receiving} · 完成 ${c.complete}`;}
    const sec=v=>v==null?'—':v.toFixed(2),states={...burstStates,waiting_headers:'等待响应头'};
    const rows=[...f.timeline].sort((a,b)=>(a.end_seconds!=null)-(b.end_seconds!=null)||b.start_seconds-a.start_seconds).slice(0,30);
    $('fader-rows').innerHTML=rows.length?rows.map(r=>`<tr><td>${esc(r.label)} · ${profiles[r.profile]}</td><td>${esc(states[r.state]||r.state)}</td><td>${r.http_status||'—'}</td><td>${sec(r.wait_seconds)}</td><td>${sec(r.start_seconds)}</td><td>${sec(r.first_output_seconds)}</td><td>${number(r.output_chars)}</td></tr>`).join(''):'<tr><td colspan="7" class="table-empty">推高任一路，观察新请求。</td></tr>';
    $('fader-window').textContent=`优先显示在途与最新请求 · 当前 ${rows.length} 条 / 累计 ${f.issued_requests} 条`;
    const actions={start:'启动',adjust:'调整',pause:'暂停补发',duration:'到时收尾',request_cap:'达到请求上限',stopped:'停止',finished:'结束'};
    $('fader-events').innerHTML=f.events.slice(-6).reverse().map(e=>`<div>${sec(e.at_seconds)}s · ${actions[e.action]||esc(e.action)} · 短 ${e.targets[0]} / 中 ${e.targets[1]} / 长 ${e.targets[2]}${e.paused?' · 暂停中':''}</div>`).join('');
    $('download-fader-events').hidden=!job.downloads.includes('fader-events.json');$('download-fader-events').href=`/api/jobs/${job.id}/fader-events.json`;
    sync();
  }
  return {init,sync,render,config,isMode};
})();

File: relay_lab/faders.py
"""Live three-channel HTTP load. Waiting is client evidence, not a server queue claim."""
import asyncio
import copy
import threading
import time
from collections import Counter

from .metrics import stage_summary
from .occupancy import Occupancy
from .report import atomic_json
from .resources import Monitor

PROFILES = ('short', 'medium', 'long')


def validate_targets(targets, ceiling):
    if (not isinstance(targets, list) or len(targets) != 3
            or any(type(n) is not int or not 0 <= n <= ceiling for n in targets)
            or sum(targets) > ceiling):
        raise ValueError('请将三个推子的总目标设在本次客户端并发上限以内。')
    return list(targets)


class Faders:
    def __init__(self, config, output, *, record=True):
        self.cfg = copy.deepcopy(config)
        self.output = output
        self.lock = threading.RLock()
        self.targets = list(config['targets'])
        self.paused = False
        self.accepting = True
        self.reason = None
        self.started = time.time()
        self.finished = None
        self.rows = {}
        self.events = []
        self.version = 0
        if record:
            self._event('start')

    def _event(self, action):
        self.version += 1
        self.events.append({'version': self.version, 'at_seconds': time.time() - self.started,
                            'action': action, 'targets': list(self.targets), 'paused': self.paused})
        atomic_json(self.output / 'fader-events.json', {'events': self.events})

    def adjust(self, body):
        with self.lock:
            if not self.accepting:
                raise ValueError('请等待本次测试收尾完成后新建测试。')
            if not isinstance(body, dict) or set(body) - {'targets', 'paused'} or not body:
                raise ValueError('请提供有效的推子目标或补发状态。')
            targets = validate_targets(body.get('targets', self.targets), self.cfg['max_inflight'])
            paused = body.get('paused', self.paused)
            if type(paused) is not bool:
                raise ValueError('请提供有效的补发状态。')
            if len(self.events) >= 10000:
                raise ValueError('请结束当前测试后新建测试，调节记录已达到上限。')
            if targets != self.targets or paused != self.paused:
                self.targets, self.paused = targets, paused
                self._event('pause' if paused else 'adjust')
            return self.version

    def desired(self):
        with self.lock:
            return ([0, 0, 0] if self.paused or not self.accepting else list(self.targets)), self.version

    def close(self, reason):
        with self.lock:
            if self.accepting:
                self.accepting = False
                self.reason = reason
                self._event(reason)

    def track(self, result):
        with self.lock:
            self.rows[result.request_id] = result.public()

    def snapshot(self):
        with self.lock:
            now = self.finished or time.time()
            rows = list(self.rows.values())
            pending = [r for r in rows if r['ended_at'] is None]
            ended = sorted((r for r in rows if r['ended_at'] is not None), key=lambda r: r['ended_at'])
            visible = sorted(pending + ended[-60:], key=lambda r: r['started_at'], reverse=True)
            timeline = []
            for r in visible:
                end = r['ended_at'] or now
                state = ('complete' if r['success'] else r['error']) if r['ended_at'] else (
                    'receiving' if r['first_content_at'] else 'waiting_first_output' if r['headers_at'] else 'waiting_headers')
                rel = lambda t: t - self.started if t is not None else None
                timeline.append({'label': r['burst_label'], 'profile': r['workload_profile'], 'state': state,
                                 'http_status': r['status'] or None, 'output_chars': r['output_units'],
                                 'output_limit': r['output_limit'], 'upstream_error_kind': r['upstream_error_kind'],
                                 'start_seconds': rel(r['started_at']), 'headers_seconds': rel(r['headers_at']),
                                 'first_output_seconds': rel(r['first_content_at']), 'end_seconds': rel(r['ended_at']),
                                 'wait_seconds': (r['first_content_at'] or end) - r['started_at'],
                                 'client_gate_ms': r['queue_ms'],
                                 'silent_seconds': end - r['last_output_at'] if r['last_output_at'] else None})
            channels = []
            for profile, target in zip(PROFILES, self.targets):
                group = [r for r in rows if r['workload_profile'] == profile]
                live = [r for r in group if r['ended_at'] is None]
                channels.append({'profile': profile, 'target': target, 'issued': len(group), 'inflight': len(live),
                                 'waiting': sum(r['first_content_at'] is None for r in live),
                                 'receiving': sum(r['first_content_at'] is not None for r in live),
                                 'headers_received': sum(r['headers_at'] is not None for r in group),
                                 'received_output': sum(r['first_content_at'] is not None for r in group),
                                 'complete': sum(r['success'] for r in group),
                                 'errors': dict(Counter(r['error'] for r in group if r['error']))})
            return {'load_mode': 'faders', 'targets': list(self.targets), 'paused': self.paused,
                    'accepting': self.accepting, 'stop_reason': self.reason, 'version': self.version,
                    'max_inflight': self.cfg['max_inflight'], 'max_requests': self.cfg['max_requests'],
                    'duration': self.cfg['duration'], 'refill_interval': self.cfg['refill_interval'],
                    'output_limits': list(self.cfg['output_limits']), 'elapsed_seconds': now - self.started,
                    'issued_requests': len(rows), 'inflight': len(pending), 'channels': channels,
                    'events': copy.deepcopy(self.events), 'timeline': timeline,
                    'timeline_scope': 'all_pending_and_latest_60_ended; all_requests_in_results_jsonl',
                    'server_queue_verified': False, 'upstream_account_occupancy_verified': False}


async def run(lab):
    control, cfg = lab.faders, lab.cfg['faders']
    rows, tasks = [], {}
    serial = [0, 0, 0]
    issued = 0
    started = time.monotonic()
    next_refill = started
    seen_version = 0
    label = 'live-faders'
    async with lab.target() as (adapter, mock):
        occupancy = lab.live_occupancy = adapter.occupancy = Occupancy(label, sum(cfg['targets']))
        occupancy.set_target(sum(cfg['targets']))
        adapter.on_progress = control.track
        lab.current_phase = 'faders'
        monitor = Monitor(adapter, mock is not None)
        monitoring = asyncio.create_task(monitor.run())

        async def worker(index, name, concurrency):
            options = ({'chunks': 40, 'chunk_delay': cfg['mock_durations'][index] / 40} if adapter.is_mock else {})
            result = await adapter.request(label, max(1, concurrency), workload=PROFILES[index],
                                           output_tokens=cfg['output_limits'][index],
                                           limit_field=lab.cfg['workload']['limit_field'], burst_label=name,
                                           account=0 if adapter.is_mock else None, **options)
            rows.append(result)
            lab.artifacts.append(result)

        try:
            while True:
                now = time.monotonic()
                for task in list(tasks):
                    if task.done():
                        del tasks[task]
                        task.result()
                if lab.stop.is_set():
                    control.close('stopped')
                    occupancy.close_load('stopped')
                    break
                if now - started >= cfg['duration'] or issued >= cfg['max_requests']:
                    reason = 'duration' if now - started >= cfg['duration'] else 'request_cap'
                    control.close(reason)
                    occupancy.close_load(reason)
                    lab.current_phase = 'draining'
                desired, version = control.desired()
                if version != seen_version:
                    occupancy.set_target(sum(desired))
                    next_refill, seen_version = now, version
                if not control.accepting and not tasks:
                    break
                if control.accepting and now >= next_refill:
                    counts = Counter(tasks.values())
                    needs = [max(0, desired[i] - counts[i]) for i in range(3)]
                    budget = min(cfg['max_inflight'] - len(tasks), cfg['max_requests'] - issued)
                    # Interleave channels in each refill wave. No hidden small client connection pool.
                    for n in range(max(needs)):
                        for i in range(3):
                            if n >= needs[i] or budget <= 0:
                                continue
                            serial[i] += 1
                            task = asyncio.create_task(worker(i, 'SML'[i] + str(serial[i]), sum(desired)))
                            tasks[task] = i
                            issued += 1
                            budget -= 1
                    next_refill = now + cfg['refill_interval']
                try:
                    await asyncio.wait_for(lab.stop.wait(), .05)
                except TimeoutError:
                    pass
        finally:
            control.close('stopped' if lab.stop.is_set() else 'finished')
            for task in tasks:
                if not task.done():
                    task.cancel()
            await asyncio.gather(*tasks, return_exceptions=True)
            monitor.running = False
            await monitoring
            occupancy.finish(control.reason)
            adapter.on_progress = None
            control.finished = time.time()
        stage = stage_summary(label, max(1, occupancy.peak), rows, time.monotonic() - started, lab.cfg['min_samples'])
        stage.update({'load_mode': 'faders', 'workload_profile': 'mixed', 'output_limit': None,
                      'occupancy': occupancy.snapshot(), 'resources': monitor.summary(),
                      'max_observed_inflight': mock.peak_active if mock else adapter.peak_active})
        lab.stages.append(stage)
    return {'max_stable_concurrency': None, 'capacity_limit_determined': False, 'low_confidence': True,
            'concurrency_scope': 'client_requests_only', 'faders': control.snapshot()}
